graphql/schemabuilder/reflect.go

```diff
@@ -315,63 +325,74 @@ func wrapPtrParser(inner *argParser) *argParser {
 }
 
 func (sb *schemaBuilder) makeStructParser(typ reflect.Type) (*argParser, graphql.Type, error) {
-	fields := make(map[string]argField)
-	argType := &graphql.InputObject{
-		Name:        typ.Name(),
-		InputFields: make(map[string]graphql.Type),
-	}
-	if argType.Name != "" {
-		argType.Name += "_InputObject"
-	}
-
-	if typ.Kind() != reflect.Struct {
-		return nil, nil, fmt.Errorf("expected struct but received type %s", typ.Name())
-	}
-
-	for i := 0; i < typ.NumField(); i++ {
-		field := typ.Field(i)
-		if field.PkgPath != "" && !field.Anonymous {
-			continue
-		}
-		if field.Anonymous {
-			return nil, nil, fmt.Errorf("bad arg type %s: anonymous fields not supported", typ)
+	var fields map[string]argField
+	var argType *graphql.InputObject
+	cached, ok := typeCache[typ]
+	if ok {
+		fields = cached.fields
+		argType = cached.argType
+	} else {
```

Comment:
i wish this didn't need to introduce a new level indent. could you split this code out into a helper function and use an early return?

Reply:
refactored